modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleService.java

```diff
@@ -310,7 +330,168 @@ void run(ClusterState state) {
                     e
                 );
             }
+
+            try {
+                maybeExecuteDownsampling(state, dataStream, getTargetIndices(dataStream, indicesToExcludeForRemainingRun));
+            } catch (Exception e) {
+                logger.error(
+                    () -> String.format(
+                        Locale.ROOT,
+                        "Data stream lifecycle failed to execute downsampling for data stream [%s]",
+                        dataStream.getName()
+                    ),
+                    e
+                );
+            }
+        }
+    }
+
+    Set<Index> maybeExecuteDownsampling(ClusterState state, DataStream dataStream, List<Index> targetIndices) {
+        Set<Index> affectedIndices = new HashSet<>();
+        Metadata metadata = state.metadata();
+        for (Index index : targetIndices) {
+            if (dataStream.isIndexManagedByDataStreamLifecycle(index, metadata::index)) {
+                IndexMetadata backingIndex = metadata.index(index);
+                assert backingIndex != null : "the data stream backing indices must exist";
+                List<DataStreamLifecycle.Downsampling.Round> downsamplingRounds = dataStream.getDownsamplingRoundsFor(
+                    index,
+                    metadata::index,
+                    nowSupplier
+                );
+                if (downsamplingRounds.isEmpty()) {
+                    continue;
+                }
+
+                String indexName = index.getName();
+                IndexMetadata.DownsampleTaskStatus backingIndexDownsamplingStatus = INDEX_DOWNSAMPLE_STATUS.get(backingIndex.getSettings());
+                String backingIndexDownsamplingSource = backingIndex.getSettings().get(IndexMetadata.INDEX_DOWNSAMPLE_SOURCE_NAME_KEY);
+
+                // if the current index is not a downsample we want to mark the index as read-only before proceeding with downsampling
+                if (org.elasticsearch.common.Strings.hasText(backingIndexDownsamplingSource) == false
+                    && state.blocks().indexBlocked(ClusterBlockLevel.WRITE, indexName) == false) {
+                    // time to downsample, but first we have to mark the index as read-only
+                    affectedIndices.add(index);
+                    AddIndexBlockRequest addIndexBlockRequest = new AddIndexBlockRequest(WRITE, indexName).masterNodeTimeout(
+                        TimeValue.MAX_VALUE
+                    );
```

Comment:
Let's say this hangs, if we have the timeout as forever, then in 5 minutes we issue another request but it's de-duplicated and thus never retried. Should this at least set the timeout to the DSL interval in order to avoid this situation?

Reply:
Can you elaborate a bit on what do you mean by "this hangs"?

The master node timeout is meant to be the time we wait for the master to become available and process the request (AFAIK). So the intent here is to allow the master to process the request (as much as it needs to) without the service piling up more requests.

On a master failover, we clear the dedups so the DSL service will retry.